Add a vlModal directive and service to control vlModals

kadamwhite opened the pull request:
See #77, #82

This PR adds a vl-modal directive, and creates a VlModals service through which an individual vl-modal can be opened or closed by ID. This removes some of the scope variables needed by the existing modal solution (for the bookmarks modal), and encapsulates the modal's visibility inside the vl-modal's isolate scope.

Making this change to the bookmarks sets us up to use it for the dataset selection modal: because the bookmarks scope situation is slightly more straightforward, it makes for a good testbed.
- [x] **Still TODO** before this merges are the unit tests, which I have broken due to removing the visibility logic from the bookmarks directive.

This PR accompanies [vega/voyager#192](https://github.com/vega/voyager/pull/192) and the two should be merged in tandem in order for voyager's bookmarks to continue working.

kadamwhite committed:
Add a vlModal directive and service to control vlModals

See #77, #82

This PR adds a vl-modal directive, and creates a VlModals service through
which an individual vl-modal can be opened or closed by ID. This removes
some of the scope variables needed by the existing modal solution (for
the bookmarks modal), and encapsulates the modal's visibility inside
the vl-modal's isolate scope.

Making this change to the bookmarks sets us up to use it for the dataset
selection modal: because the bookmarks scope situation is slightly more
straightforward, it makes for a good testbed.

Still TODO before this merges are the unit tests, which I have broken
due to removing the visibility logic from the bookmarks directive.

This PR accompanies vega/voyager#192 and the two should be merged in
tandem in order for voyager's bookmarks to continue working.

kadamwhite committed:
Move logging for bookmarks close from voyager to vlui

kadamwhite commented on src/bookmarklist/bookmarklist.html:
```diff
@@ -1,8 +1,5 @@
-      ng-if="active"
-      >
+
+    <div class="vflex full-width full-height" id="evs">
```
I'll normalize this indentation outside of this PR to keep the diff clean

domoritz commented on src/index.scss:
```diff
@@ -405,21 +405,30 @@ h4 {
   background: rgba(0,0,0,0.5);
 
-  .wrapper {
+  .modal-wrapper {
```
move this to a scss file within the modal

domoritz commented on src/modal/vlmodal.js:
```diff
@@ -0,0 +1,35 @@
+ * # vlModal
+ */
+angular.module('vlui')
+  .directive('vlModal', function (VlModals) {
```
any reason not to call this `Modal`?

kadamwhite replied:
This is a good question for you and @kanitw actually, @domoritz — I did not use `<modal>` here because that is a common name and it would preclude us or others using this directive inside another app that defined its own `<modal>`, and because I had seen `<vl-plot-group>` and other vl-prefixed directives as prior art.

If we do not need this to interoperate with other angular UI code, and other potential app environments, I'm fine renaming it to `<modal>`—but there's also an argument to be made to keep the `vl-` prefixing for consistency. Let me know which way you'd like to go.

domoritz commented on src/modal/vlmodals.service.js:
```diff
@@ -0,0 +1,56 @@
+      open: function(id) {
+        var modalScope = modalsCache.get(id);
+        if (!modalScope) {
+          console.error('Unknown modal id ' + id);
```
Unregistered?

kadamwhite replied:
Will update!

kadamwhite committed:
Improve code per PR feedback and fix broken bookmarks test

The bookmarks test was only validating the modal behavior, so
I have changed it to test that the bookmarks directive requires
a parent vlModal directive: we should still add tests for the
bookmarks list behavior when we can

kadamwhite committed:
Add basic tests for vlModal directive

kadamwhite commented:
@domoritz @kanitw Other than the outstanding question about the directive name, I think this is ready to go—I have pulled the WIP commits on using this for the data load modal into a different branch that I will PR separately.

Edit: name issue has been resolved, changint to `<modal>` per discussion

kadamwhite committed:
Rename <vl-modal> directive to <modal>

domoritz commented:
@kadamwhite We usually remove the WIP label when we are done.
